Biz/Model/UserLoginBaseForm.pm

```diff
@@ -130,11 +184,26 @@ sub validate_login {
     return $realm;
 }
 
+sub _record_login_attempt {
+    my($self, $owner, $success) = @_;
+    return $self->req->with_realm($owner, sub {
+        return $self->new_other('LoginAttempt')->create({
+            login_attempt_state => $success ? $_LAS->SUCCESS : $_LAS->FAILURE,
+        });
+    });
+}
+
 sub _validate {
     my($self) = @_;
     my($owner) = $self->validate_login;
     return
         if !$owner || ($self->in_error && !$owner->require_otp);
+    if ($self->req->with_realm($owner, sub {
```

Comment:
Don't switch realms. This is expensive and complicated. This is a login form so we don't want to "go into" the realm until we know we have authenticated the user. In this case, you would unauth_load_last. Also this type of code is better modularized in another sub so this could would look like:

```perl
return
    unless _validate_login_attempt($self);
```